Godot compilation failed on Windows

**Operating system or device, Godot version, GPU Model and driver (if graphics related):**
Windows 10, Godot 3.0 alpha
After PR https://github.com/godotengine/godot/pull/11518

**Issue description:**
Godot doesnt compiles under windows with the following trace:
```
scons platform=windows
scons: Reading SConscript files ...
Detected MSVC compiler: amd64
Compiled program architecture will be a 64 bit executable (forcing bits=64).
cdir is: D:\work\_misc\godot
Checking for C header file mntent.h... (cached) no
scons: done reading SConscript files.
scons: Building targets ...
[  2%] [94mCompiling [95m==> [93mplatform\windows\godot_win.cpp[0m
godot_win.cpp
d:\work\_misc\godot\core\object.h(386): error C2059: syntax error: 'constant'
d:\work\_misc\godot\core\object.h(389): error C2143: syntax error: missing ';' before '}'
d:\work\_misc\godot\core\object.h(389): error C2238: unexpected token(s) preceding ';'
d:\work\_misc\godot\core\object.h(410): error C2059: syntax error: 'private'
d:\work\_misc\godot\core\object.h(416): error C2255: 'friend': not allowed outside of a class definition
d:\work\_misc\godot\core\object.h(418): error C2255: 'friend': not allowed outside of a class definition
d:\work\_misc\godot\core\object.h(419): error C2255: 'friend': not allowed outside of a class definition
d:\work\_misc\godot\core\object.h(469): error C2178: '_class_name' cannot be declared with 'mutable' specifier
d:\work\_misc\godot\core\object.h(470): error C2178: '_class_ptr' cannot be declared with 'mutable' specifier
d:\work\_misc\godot\core\object.h(473): error C2270: '_has_user_signal': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(475): error C2270: '_get_signal_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(476): error C2270: '_get_signal_connection_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(477): error C2270: '_get_incoming_connections': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(479): error C2270: '_get_bind': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(485): error C2059: syntax error: 'protected'
d:\work\_misc\godot\core\object.h(486): error C2575: '_initialize_classv': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(486): error C3861: 'initialize_class': identifier not found
d:\work\_misc\godot\core\object.h(487): error C2575: '_setv': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(488): error C2270: '_getv': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(489): error C2270: '_get_property_listv': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(490): error C2575: '_notificationv': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(495): error C2270: '_get': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(496): error C2270: '_get_property_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(500): error C2039: '_bind_methods': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(502): error C2270: '_get_get': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(503): error C2039: '_get': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(503): error C2440: 'return': cannot convert from 'bool (__cdecl *)(const StringName &,Variant &)' to 'bool (__cdecl Object::* )(const StringName &,Variant &) const'
d:\work\_misc\godot\core\object.h(503): note: There is no context in which this conversion is possible
d:\work\_misc\godot\core\object.h(505): error C2270: '_get_set': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(506): error C2039: '_set': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(506): error C2440: 'return': cannot convert from 'bool (__cdecl *)(const StringName &,const Variant &)' to 'bool (__cdecl Object::* )(const StringName &,const Variant &)'
d:\work\_misc\godot\core\object.h(506): note: There is no context in which this conversion is possible
d:\work\_misc\godot\core\object.h(508): error C2270: '_get_get_property_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(509): error C2039: '_get_property_list': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(509): error C2440: 'return': cannot convert from 'void (__cdecl *)(List<PropertyInfo,DefaultAllocator> *)' to 'void (__cdecl Object::* )(List<PropertyInfo,DefaultAllocator> *)
const'
d:\work\_misc\godot\core\object.h(509): note: There is no context in which this conversion is possible
d:\work\_misc\godot\core\object.h(511): error C2270: '_get_notification': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(512): error C2039: '_notification': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(512): error C2440: 'return': cannot convert from 'void (__cdecl *)(int)' to 'void (__cdecl Object::* )(int)'
d:\work\_misc\godot\core\object.h(512): note: There is no context in which this conversion is possible
d:\work\_misc\godot\core\object.h(519): error C2575: '_changed_callback': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(527): error C2270: '_get_class_namev': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(529): error C3861: 'get_class_static': identifier not found
d:\work\_misc\godot\core\object.h(533): error C2270: '_get_meta_list_bind': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(534): error C2270: '_get_property_list_bind': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(535): error C2270: '_get_method_list_bind': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(539): error C2255: 'friend': not allowed outside of a class definition
d:\work\_misc\godot\core\object.h(540): error C2270: '_validate_property': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(542): error C2059: syntax error: 'public'
d:\work\_misc\godot\core\object.h(546): error C2059: syntax error: 'public'
d:\work\_misc\godot\core\object.h(551): error C2039: '_changed_callback': is not a member of 'Object'
d:\work\_misc\godot\core\object.h(382): note: see declaration of 'Object'
d:\work\_misc\godot\core\object.h(551): error C2355: 'this': can only be referenced inside non-static member functions or non-static data member initializers
d:\work\_misc\godot\core\object.h(561): error C2270: '_is_gpl_reversed': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(563): error C2270: 'get_instance_id': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(610): error C2270: 'get_class': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(611): error C2270: 'get_save_class': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(613): error C2270: 'is_class': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(614): error C2270: 'is_class_ptr': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(616): error C2270: 'get_class_name': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(629): error C2270: 'get': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(631): error C2270: 'get_property_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(633): error C2270: 'has_method': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(634): error C2270: 'get_method_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(636): error C2575: 'call': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(637): error C2575: 'call_multilevel': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(638): error C2575: 'call_multilevel_reversed': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(645): error C2270: 'getvar': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(646): error C2575: 'setvar': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(651): error C2270: 'get_script': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(655): error C2270: 'has_meta': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(657): error C2270: 'get_meta': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(658): error C2270: 'get_meta_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(662): error C2270: 'is_edited': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(663): error C2270: 'get_edited_version': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(667): error C2270: 'get_script_instance': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(674): error C2270: 'get_signal_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(675): error C2270: 'get_signal_connection_list': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(676): error C2270: 'get_all_signal_connections': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(677): error C2270: 'has_persistent_signal_connections': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(678): error C2270: 'get_signals_connected_to_this': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(682): error C2270: 'is_connected': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(687): error C2270: 'is_blocking_signals': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(689): error C2270: 'get_static_property_type': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(691): error C2270: 'get_translatable_strings': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(693): error C2270: 'get_argument_options': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(695): error C2270: 'tr': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(698): error C2270: 'is_queued_for_deletion': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(701): error C2270: 'can_translate_messages': modifiers not allowed on nonmember functions
d:\work\_misc\godot\core\object.h(713): error C2059: syntax error: ')'
d:\work\_misc\godot\core\object.h(714): error C2588: '::~Object': illegal global destructor
d:\work\_misc\godot\core\object.h(714): error C2575: 'Object': only member functions and bases can be virtual
d:\work\_misc\godot\core\object.h(714): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
d:\work\_misc\godot\core\object.h(715): error C2059: syntax error: '}'
d:\work\_misc\godot\core\object.h(715): error C2143: syntax error: missing ';' before '}'
d:\work\_misc\godot\core\object.h(717): error C2065: 'p_object': undeclared identifier
d:\work\_misc\godot\core\object.h(717): error C2365: 'predelete_handler': redefinition; previous definition was 'function'
d:\work\_misc\godot\core\object.h(718): error C2065: 'p_object': undeclared identifier
d:\work\_misc\godot\core\object.h(718): error C2182: 'postinitialize_handler': illegal use of type 'void'
d:\work\_misc\godot\core\object.h(718): error C2365: 'postinitialize_handler': redefinition; previous definition was 'function'
d:\work\_misc\godot\core\object.h(724): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
d:\work\_misc\godot\core\object.h(724): error C2143: syntax error: missing ',' before '*'
d:\work\_misc\godot\core\object.h(735): error C2059: syntax error: '>'
d:\work\_misc\godot\core\object.h(735): error C2976: 'HashMap': too few template arguments
d:\work\_misc\godot\core\hash_map.h(58): note: see declaration of 'HashMap'
d:\work\_misc\godot\core\object.h(736): error C2059: syntax error: ','
d:\work\_misc\godot\core\object.h(744): error C2061: syntax error: identifier 'Object'
d:\work\_misc\godot\core\object.h(745): error C2061: syntax error: identifier 'Object'
d:\work\_misc\godot\core\object.h(750): error C2061: syntax error: identifier 'Object'
d:\work\_misc\godot\core\object.h(750): fatal error C1003: error count exceeds 100; stopping compilation
scons: *** [platform\windows\godot_win.windows.tools.64.obj] Error 2
scons: building terminated because of errors.
```


**Steps to reproduce:**
Try build godot for windows under MSVC

--- a/core/safe_refcount.cpp
+++ b/core/safe_refcount.cpp
@@ -1,0 +1,169 @@
+/*************************************************************************/
+/*  safe_refcount.cpp                                                    */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2017 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2017 Godot Engine contributors (cf. AUTHORS.md)    */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "safe_refcount.h"
+
+#if defined(_MSC_VER)
+
+/* Implementation for MSVC-Windows */
+
+// don't pollute my namespace!
+#include <windows.h>
+
+#define ATOMIC_CONDITIONAL_INCREMENT_BODY(m_pw, m_win_type, m_win_cmpxchg, m_cpp_type) \
+	/* try to increment until it actually works */                                     \
+	/* taken from boost */                                                             \
+	while (true) {                                                                     \
+		m_cpp_type tmp = static_cast<m_cpp_type const volatile &>(*(m_pw));            \
+		if (tmp == 0)                                                                  \
+			return 0; /* if zero, can't add to it anymore */                           \
+		if (m_win_cmpxchg((m_win_type volatile *)(m_pw), tmp + 1, tmp) == tmp)         \
+			return tmp + 1;                                                            \
+	}
+
+#define ATOMIC_EXCHANGE_IF_GREATER_BODY(m_pw, m_val, m_win_type, m_win_cmpxchg, m_cpp_type) \
+	while (true) {                                                                          \
+		m_cpp_type tmp = static_cast<m_cpp_type const volatile &>(*(m_pw));                 \
+		if (tmp >= m_val)                                                                   \
+			return tmp; /* already greater, or equal */                                     \
+		if (m_win_cmpxchg((m_win_type volatile *)(m_pw), m_val, tmp) == tmp)                \
+			return m_val;                                                                   \
+	}
+
+_ALWAYS_INLINE_ uint32_t _atomic_conditional_increment_impl(register uint32_t *pw) {
+
+	ATOMIC_CONDITIONAL_INCREMENT_BODY(pw, LONG, InterlockedCompareExchange, uint32_t)
+}
+
+_ALWAYS_INLINE_ uint32_t _atomic_decrement_impl(register uint32_t *pw) {
+
+	return InterlockedDecrement((LONG volatile *)pw);
+}
+
+_ALWAYS_INLINE_ uint32_t _atomic_increment_impl(register uint32_t *pw) {
+
+	return InterlockedIncrement((LONG volatile *)pw);
+}
+
+_ALWAYS_INLINE_ uint32_t _atomic_sub_impl(register uint32_t *pw, register uint32_t val) {
+
+	return InterlockedExchangeAdd((LONG volatile *)pw, -(int32_t)val) - val;
+}
+
+_ALWAYS_INLINE_ uint32_t _atomic_add_impl(register uint32_t *pw, register uint32_t val) {
+
+	return InterlockedAdd((LONG volatile *)pw, val);
+}
+
+_ALWAYS_INLINE_ uint32_t _atomic_exchange_if_greater_impl(register uint32_t *pw, register uint32_t val) {
+
+	ATOMIC_EXCHANGE_IF_GREATER_BODY(pw, val, LONG, InterlockedCompareExchange, uint32_t)
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_conditional_increment_impl(register uint64_t *pw) {
+
+	ATOMIC_CONDITIONAL_INCREMENT_BODY(pw, LONGLONG, InterlockedCompareExchange64, uint64_t)
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_decrement_impl(register uint64_t *pw) {
+
+	return InterlockedDecrement64((LONGLONG volatile *)pw);
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_increment_impl(register uint64_t *pw) {
+
+	return InterlockedIncrement64((LONGLONG volatile *)pw);
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_sub_impl(register uint64_t *pw, register uint64_t val) {
+
+	return InterlockedExchangeAdd64((LONGLONG volatile *)pw, -(int64_t)val) - val;
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_add_impl(register uint64_t *pw, register uint64_t val) {
+
+	return InterlockedAdd64((LONGLONG volatile *)pw, val);
+}
+
+_ALWAYS_INLINE_ uint64_t _atomic_exchange_if_greater_impl(register uint64_t *pw, register uint64_t val) {
+
+	ATOMIC_EXCHANGE_IF_GREATER_BODY(pw, val, LONGLONG, InterlockedCompareExchange64, uint64_t)
+}
+
+// The actual advertised functions; they'll call the right implementation
+
+uint32_t atomic_conditional_increment(register uint32_t *counter) {
+	return _atomic_conditional_increment_impl(counter);
+}
+
+uint32_t atomic_decrement(register uint32_t *pw) {
+	return _atomic_decrement_impl(pw);
+}
+
+uint32_t atomic_increment(register uint32_t *pw) {
+	return _atomic_increment_impl(pw);
+}
+
+uint32_t atomic_sub(register uint32_t *pw, register uint32_t val) {
+	return _atomic_sub_impl(pw, val);
+}
+
+uint32_t atomic_add(register uint32_t *pw, register uint32_t val) {
+	return _atomic_add_impl(pw, val);
+}
+
+uint32_t atomic_exchange_if_greater(register uint32_t *pw, register uint32_t val) {
+	return _atomic_exchange_if_greater_impl(pw, val);
+}
+
+uint64_t atomic_conditional_increment(register uint64_t *counter) {
+	return _atomic_conditional_increment_impl(counter);
+}
+
+uint64_t atomic_decrement(register uint64_t *pw) {
+	return _atomic_decrement_impl(pw);
+}
+
+uint64_t atomic_increment(register uint64_t *pw) {
+	return _atomic_increment_impl(pw);
+}
+
+uint64_t atomic_sub(register uint64_t *pw, register uint64_t val) {
+	return _atomic_sub_impl(pw, val);
+}
+
+uint64_t atomic_add(register uint64_t *pw, register uint64_t val) {
+	return _atomic_add_impl(pw, val);
+}
+
+uint64_t atomic_exchange_if_greater(register uint64_t *pw, register uint64_t val) {
+	return _atomic_exchange_if_greater_impl(pw, val);
+}
+#endif

--- a/core/safe_refcount.h
+++ b/core/safe_refcount.h
@@ -150,97 +150,25 @@ static _ALWAYS_INLINE_ T atomic_exchange_if_greater(register T *pw, register V v
 }
 
 #elif defined(_MSC_VER)
+// For MSVC use a separate compilation unit to prevent windows.h from polluting
+// the global namespace.
+uint32_t atomic_conditional_increment(register uint32_t *pw);
+uint32_t atomic_decrement(register uint32_t *pw);
+uint32_t atomic_increment(register uint32_t *pw);
+uint32_t atomic_sub(register uint32_t *pw, register uint32_t val);
+uint32_t atomic_add(register uint32_t *pw, register uint32_t val);
+uint32_t atomic_exchange_if_greater(register uint32_t *pw, register uint32_t val);
 
-/* Implementation for MSVC-Windows */
-
-// don't pollute my namespace!
-#include <windows.h>
-
-#define ATOMIC_CONDITIONAL_INCREMENT_BODY(m_pw, m_win_type, m_win_cmpxchg, m_cpp_type) \
-	/* try to increment until it actually works */                                     \
-	/* taken from boost */                                                             \
-	while (true) {                                                                     \
-		m_cpp_type tmp = static_cast<m_cpp_type const volatile &>(*(m_pw));            \
-		if (tmp == 0)                                                                  \
-			return 0; /* if zero, can't add to it anymore */                           \
-		if (m_win_cmpxchg((m_win_type volatile *)(m_pw), tmp + 1, tmp) == tmp)         \
-			return tmp + 1;                                                            \
-	}
-
-#define ATOMIC_EXCHANGE_IF_GREATER_BODY(m_pw, m_val, m_win_type, m_win_cmpxchg, m_cpp_type) \
-	while (true) {                                                                          \
-		m_cpp_type tmp = static_cast<m_cpp_type const volatile &>(*(m_pw));                 \
-		if (tmp >= m_val)                                                                   \
-			return tmp; /* already greater, or equal */                                     \
-		if (m_win_cmpxchg((m_win_type volatile *)(m_pw), m_val, tmp) == tmp)                \
-			return m_val;                                                                   \
-	}
-
-static _ALWAYS_INLINE_ uint32_t atomic_conditional_increment(register uint32_t *pw) {
-
-	ATOMIC_CONDITIONAL_INCREMENT_BODY(pw, LONG, InterlockedCompareExchange, uint32_t)
-}
-
-static _ALWAYS_INLINE_ uint32_t atomic_decrement(register uint32_t *pw) {
-
-	return InterlockedDecrement((LONG volatile *)pw);
-}
-
-static _ALWAYS_INLINE_ uint32_t atomic_increment(register uint32_t *pw) {
-
-	return InterlockedIncrement((LONG volatile *)pw);
-}
-
-static _ALWAYS_INLINE_ uint32_t atomic_sub(register uint32_t *pw, register uint32_t val) {
-
-	return InterlockedExchangeAdd((LONG volatile *)pw, -(int32_t)val) - val;
-}
-
-static _ALWAYS_INLINE_ uint32_t atomic_add(register uint32_t *pw, register uint32_t val) {
-
-	return InterlockedAdd((LONG volatile *)pw, val);
-}
-
-static _ALWAYS_INLINE_ uint32_t atomic_exchange_if_greater(register uint32_t *pw, register uint32_t val) {
-
-	ATOMIC_EXCHANGE_IF_GREATER_BODY(pw, val, LONG, InterlockedCompareExchange, uint32_t)
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_conditional_increment(register uint64_t *pw) {
-
-	ATOMIC_CONDITIONAL_INCREMENT_BODY(pw, LONGLONG, InterlockedCompareExchange64, uint64_t)
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_decrement(register uint64_t *pw) {
-
-	return InterlockedDecrement64((LONGLONG volatile *)pw);
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_increment(register uint64_t *pw) {
-
-	return InterlockedIncrement64((LONGLONG volatile *)pw);
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_sub(register uint64_t *pw, register uint64_t val) {
-
-	return InterlockedExchangeAdd64((LONGLONG volatile *)pw, -(int64_t)val) - val;
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_add(register uint64_t *pw, register uint64_t val) {
-
-	return InterlockedAdd64((LONGLONG volatile *)pw, val);
-}
-
-static _ALWAYS_INLINE_ uint64_t atomic_exchange_if_greater(register uint64_t *pw, register uint64_t val) {
-
-	ATOMIC_EXCHANGE_IF_GREATER_BODY(pw, val, LONGLONG, InterlockedCompareExchange64, uint64_t)
-}
+uint64_t atomic_conditional_increment(register uint64_t *pw);
+uint64_t atomic_decrement(register uint64_t *pw);
+uint64_t atomic_increment(register uint64_t *pw);
+uint64_t atomic_sub(register uint64_t *pw, register uint64_t val);
+uint64_t atomic_add(register uint64_t *pw, register uint64_t val);
+uint64_t atomic_exchange_if_greater(register uint64_t *pw, register uint64_t val);
 
 #else
-
 //no threads supported?
 #error Must provide atomic functions for this platform or compiler!
-
 #endif
 
 struct SafeRefCount {

